dpgen2/exploration/report/report_adaptive_lower.py

```diff
@@ -0,0 +1,330 @@
+            self.level_f_lo = self.level_f_hi
+        else:
+            self.level_f_lo = coll_f[-numb_candi_f][0]
+        # add to candidate set
+        for ii in range(len(coll_f) - numb_candi_f, len(coll_f)):
+            self.candi.add(tuple(coll_f[ii][1:]))
+        for ii in range(len(coll_v) - numb_candi_v, len(coll_v)):
+            self.candi.add(tuple(coll_v[ii][1:]))
+        # accurate set is substracted by the candidate set
+        self.accur = self.accur - self.candi
+
+
+    def _record_one_traj(
+            self,
+            tt,
+            md_f : np.ndarray,
+            md_v : Optional[np.ndarray] = None,            
+    ):
+        """
+        Record one trajctory. 
+
+        tt:             traj index
+        md_f, md_v:     model deviations of force and virial
+        """
+        # check consistency
+        if self.has_virial and md_v is None:
+            raise FatalError(
+                "report requires virial model deviation, but no virial "
+                "model deviation is provided."
+            )
+        # fake md_v as zeros if None is provided
+        if md_v is None:
+            md_v = np.zeros_like(md_f)
+        # loop over frames
+        nframes = md_f.shape[0]
+        assert nframes == md_v.shape[0]
+        failed = []
+        accur = set()
+        coll_f = []
+        coll_v = []
+        for ii in range(nframes):
+            if md_f[ii] > self.level_f_hi or md_v[ii] > self.level_v_hi:
+                failed.append((tt, ii))
+            else:
+                coll_f.append([md_f[ii], tt, ii])
+                coll_v.append([md_v[ii], tt, ii])
+                # now accur takes all non-failed frames,
+                # will be substracted by candidate later
+                accur.add((tt, ii))
+        return nframes, accur, failed, coll_f, coll_v
+
+    def _sequence_conv(
+            self,
+            seq,
+    )->bool:
+        if len(seq) <= 1:
+            return False
+        conv = [ abs(seq[ii-1] - seq[ii]) < self.conv_tolerance 
+                 for ii in range(1,len(seq)) ]
+        return all(conv)
+
+    def converged(
+            self,
+            reports,
+    )->bool:
+        if 1 + len(reports) < self.n_checked_steps:
+            return False
+        else:
+            all_level_f = [ii.level_f_lo for ii in reports] + [self.level_f_lo]
+            all_level_f = all_level_f[-self.n_checked_steps:]
+            conv = self._sequence_conv(all_level_f)
+            if self.has_virial:
+                all_level_v = [ii.level_v_lo for ii in reports] + [self.level_v_lo]
+                all_level_v = all_level_v[-self.n_checked_steps:]
+                conv = conv and self._sequence_conv(all_level_v)
+            return conv            
+
+    def failed_ratio(
+            self,
+            tag = None,
+    ):
+        return float(len(self.failed)) / float(self.nframes)
+
+    def accurate_ratio(
+            self,
+            tag = None,
```

Comment:
I don't quite understand what `tag` is used for.

Reply:
It is reserved for further use, e.g. we want to calculate the accurate ratio of a certain subgroup of the explored configurations.